src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+  // value segment.
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto values = segment_ptr->values();
+  auto n_values = values.size();
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto value = values[offset];
+    if (matches_search_value<T>(value)) {
+      include_rows_ptr->emplace_back(offset);
+    }
+  }
+  return include_rows_ptr;
+}
+
+template <typename T>
+std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_segment(
+    const std::shared_ptr<const DictionarySegment<T>> segment_ptr) const {
+  // determine which values match the filter condition in a
+  // dictionary segment.
+  // The current implementation uses an unoptimized approach where
+  // we unpack each value in the dictionary segment and check the
+  // condition. We could improve performance by making use of the
+  // ordered dictionary directly.
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto dictionary = segment_ptr->dictionary();
+  auto attribute_vector_ptr = segment_ptr->attribute_vector();
+  auto n_values = attribute_vector_ptr->size();
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto value = dictionary[attribute_vector_ptr->get(offset)];
+    if (matches_search_value<T>(value)) {
+      include_rows_ptr->emplace_back(offset);
+    }
+  }
+  return include_rows_ptr;
+}
+
+template <typename T>
+std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_segment(
+    const std::shared_ptr<const ReferenceSegment> segment_ptr) const {
+  // determine which values match the filter condition in a
+  // reference segment.
+  // We cannot determine if a row matches the filter condition
+  // based just on the information in the reference segment. We
+  // need to go to the table that the reference segment points to
+  // in order to retrieve the actual values and perform the filtering.
+  // We need to treat the reference segment differently based on whether
+  // it points to a value segment or a dictionary segment.
+
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto referenced_table_ptr = segment_ptr->referenced_table();
+  auto referenced_column_id = segment_ptr->referenced_column_id();
+  auto pos_list_ptr = segment_ptr->pos_list();
+  auto n_values = pos_list_ptr->size();
+
+  // iterate over all rows in the reference segment. Retrieve the
+  // segment that the reference segment points to and get the actual
+  // value from that segment.
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto row_id = (*pos_list_ptr)[offset];
+    auto chunk_id = row_id.chunk_id;
+    auto chunk_offset = row_id.chunk_offset;
+    auto referenced_segment_ptr = referenced_table_ptr->get_chunk(chunk_id)->get_segment(referenced_column_id);
+
+    // referenced segment is ValueSegment
+    const auto typed_value_segment_ptr = std::dynamic_pointer_cast<ValueSegment<T>>(referenced_segment_ptr);
```

Comment:
`ReferenceSegment::operator[]` returns an `AllTypeVariant` which you could use for comparison. No need to cast to `ValueSegment` here which also looks expensive inside the loop iterating over all rows.

Additionally this looks duplicated to your code inside function `TableScan::scan_segment(
     const std::shared_ptr<const ValueSegment<T>> - I think you could have called you own function here.

Reply:
I don't really know if this applies hier, since we're working on the segments of the referenced table, not on the reference segment itself.
Code duplication should be reduces, tho'